crates/js-backend/src/function.rs

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright 2023 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+use crate::JsInstance;
+use crate::JsWasmBackend;
+use crate::JsImportCallContext;
+use crate::JsContext;
+use crate::JsContextMut;
+use crate::js_conversions::js_array_from_wval_array;
+use crate::js_conversions::wval_array_from_js_array;
+use crate::js_conversions::wval_from_js;
+use crate::js_conversions::wval_to_i32;
+use crate::store::JsStoreInner;
+use crate::store::FunctionHandle;
+
+use marine_wasm_backend_traits::impl_for_each_function_signature;
+use marine_wasm_backend_traits::replace_with;
+use marine_wasm_backend_traits::prelude::*;
+
+use anyhow::anyhow;
+use js_sys::Array;
+use wasm_bindgen::prelude::*;
+
+#[derive(Clone)]
+pub struct HostImportFunction {
+    pub(crate) store_handle: FunctionHandle,
+}
+
+#[derive(Clone)]
+pub struct WasmExportFunction {
+    pub(crate) store_handle: FunctionHandle,
+
+    /// Instance this export is extracted from.
+    pub(crate) bound_instance: JsInstance,
+}
+
+pub(crate) struct StoredFunction {
+    pub(crate) js_func: js_sys::Function,
+    pub(crate) sig: FuncSig,
+}
+
+impl StoredFunction {
+    pub(crate) fn new(js_func: js_sys::Function, sig: FuncSig) -> Self {
+        Self { js_func, sig }
+    }
+}
+
+impl WasmExportFunction {
+    pub(crate) fn new_stored(
+        ctx: &mut impl AsContextMut<JsWasmBackend>,
+        instance: JsInstance,
+        func: js_sys::Function,
+        sig: FuncSig,
+    ) -> Self {
+        let handle = ctx
+            .as_context_mut()
+            .inner
+            .store_function(StoredFunction::new(func, sig));
+
+        Self {
+            store_handle: handle,
+            bound_instance: instance,
+        }
+    }
+
+    pub(crate) fn stored_mut<'store>(
+        &self,
+        ctx: JsContextMut<'store>,
+    ) -> &'store mut StoredFunction {
+        &mut ctx.inner.functions[self.store_handle]
+    }
+
+    fn call_inner(
```

Comment:
should it be public?

Reply:
No, it is only intended to be used inside of implementation of `ExportFunction::call`. Using it in other places will break wasm call stack invariant'